src/main/java/com/cloudbees/jenkins/support/startup/StartupReport.java

```diff
@@ -0,0 +1,166 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.jenkins.support.startup;
+
+import com.cloudbees.jenkins.support.SupportPlugin;
+import com.cloudbees.jenkins.support.impl.ThreadDumps;
+import com.cloudbees.jenkins.support.timer.FileListCap;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.XmlFile;
+import hudson.init.InitMilestone;
+import hudson.init.Initializer;
+import hudson.model.AsyncPeriodicWork;
+import hudson.model.Saveable;
+import hudson.model.TaskListener;
+import hudson.model.listeners.SaveableListener;
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import jenkins.util.SystemProperties;
+
+/**
+ * Builds up startup report by collecting thread dumps during startup as well as the timing for each milestone.
+ */
+@Extension
+public final class StartupReport extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(StartupReport.class.getName());
+    private static final int THREAD_DUMPS_TO_RETAIN =
+            SystemProperties.getInteger(StartupReport.class.getName() + ".THREAD_DUMPS_TO_RETAIN", 40);
+    private static final int TOTAL_ITERATIONS =
+            SystemProperties.getInteger(StartupReport.class.getName() + ".TOTAL_ITERATIONS", 1);
+    private static final int INITIAL_DELAY_SECONDS =
+            SystemProperties.getInteger(StartupReport.class.getName() + ".INITIAL_DELAY_SECONDS", 300);
+    private static final int RECURRENCE_PERIOD_SECONDS =
+            SystemProperties.getInteger(StartupReport.class.getName() + ".RECURRENCE_PERIOD_SECONDS", 30);
+    private static final int DELAY_BETWEEN_THREAD_DUMPS_MS =
+            SystemProperties.getInteger(StartupReport.class.getName() + ".DELAY_BETWEEN_THREAD_DUMPS_MS", 1000);
+
+    private final FileListCap logs =
+            new FileListCap(new File(SupportPlugin.getRootDirectory(), "startup-threaddumps"), THREAD_DUMPS_TO_RETAIN);
+
+    /**
+     * Record the timestamp at which each milestone was reached.
+     */
+    private Map<InitMilestone, Long> timesPerMilestone = new ConcurrentHashMap<>();
```

Comment:
Using `java.time` types would be more legible and less prone to confusing instants with durations.